difftest/doc/rational.md

```diff
@@ -0,0 +1,38 @@
+
+#### Objective
+The goal of offline difftest is to ensure the correctness of each instruction’s observed behavior during processor execution, while providing a flexible validation mechanism for handling undefined behavior. The validation process involves three steps:
+
+1. Running the driver to generate the DUT (Device Under Test) trace.
+2. Using the DUT trace to generate the model trace.
+3. Detecting instruction differences by comparing the DUT trace and model trace using the difftest algorithm.
+
+#### Design Choices
+
+##### 1. **DUT Trace Design and Encoding**
+   - **Design:** The DUT trace design must consider how to efficiently and accurately record the execution state of each instruction, including the execution address, register values, memory state, etc. Currently, using `printf` to output the simulation address is a preliminary solution, with plans to possibly use XDMA (high-speed data transfer interface) to extract FPGA traces in the future for higher verification efficiency.
+   - **Encoding:** Proper encoding of the trace will facilitate fast alignment and parsing in the comparison stage.
+
+##### 2. **Simulator Integration**
+   - **Using SAIL Simulator:** SAIL offers a highly formalized and customizable architecture simulation framework. Despite its slower speed, SAIL’s precision and flexibility make it well-suited for offline verification needs.
+     - **Removing Platform Dependencies:** To simplify the simulation, platform-related parts of SAIL will be removed, focusing on validating the instruction set and processor core.
+     - **Function Patch:** For example, in validating FP (floating-point) related instructions, certain functions, such as `fp reduce order`, may need to be replaced.
+     - **Handling Undefined Behavior:** To verify undefined behavior, an interface will be provided for the simulator to inject specific behavior. This may involve forking the SAIL project or directly customizing its source code.
+
+##### 3. **DiffTest Algorithm**
+   - **Instruction Alignment:** During the comparison process, it’s crucial to ensure that the DUT and model instruction streams are properly aligned. This may require special alignment algorithms to handle challenges arising from out-of-order or parallel execution.
+   - **Result Comparison:** The core of the comparison is to ensure that the execution results of each instruction match. This can be done via streaming-based real-time comparison or through other efficient differential calculation methods.
+
+#### Potential Challenges and Solutions
+
+##### 1. **Memory Behavior Issues**
+   - When memory access order differs, it may lead to processor behavior discrepancies. In such cases, alignment strategies and memory model simulation may be needed to reduce the impact of memory ordering on the behavior.
```

Comment:
```suggestion
   - When memory access order differs, it may lead to processor behavior discrepancies. In such cases, alignment strategies and memory model simulation may be needed.
```